src/Raven.Server/Documents/Indexes/AsyncWaitForIndexing.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Abstractions.Data;
+
+namespace Raven.Server.Documents.Indexes
+{
+    public class AsyncWaitForIndexing : IDisposable
+    {
+        private readonly string _indexName;
+        private readonly Stopwatch _queryDuration;
+        private readonly TimeSpan _waitTimeout;
+        private readonly DocumentsNotifications _notifications;
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
+
+        public AsyncWaitForIndexing(string indexName, Stopwatch queryDuration, TimeSpan waitTimeout, DocumentsNotifications notifications)
+        {
+            _indexName = indexName;
+            _queryDuration = queryDuration;
+            _waitTimeout = waitTimeout;
+            _notifications = notifications;
+            _notifications.OnIndexChange += OnIndexChange;
```

Comment:
Using semaphore on every call is going to be expensive, no?
We can use an AsyncManualResetEvent at the index level, which will be triggered on the end of the indexing batch.
That way, we have a single instance, and we only pay the price per indexing batch, instead of per query.
It also saves us the need to register to on index changed notifications